docs/topics/compatibility-guides/compatibility-guide-21.md

```diff
@@ -0,0 +1,479 @@
+>
+> **Deprecation cycle**:
+>
+> - 1.4.30: introduce locale-agnostic alternatives as experimental API
+> - 1.5.0: deprecate locale-sensitive case conversion functions with a warning
+> - 2.1.0: raise the warning to an error`
+
+### Remove kotlin-stdlib-common JAR artifact
+
+> **Issue**: [KT-62159](https://youtrack.jetbrains.com/issue/KT-62159)
+>
+> **Component**: kotlin-stdlib
+>
+> **Incompatible change type**: binary
+>
+> **Short summary**:  The `kotlin-stdlib-common.jar` artifact, previously used for legacy multiplatform declarations metadata, is deprecated and replaced by `.klib` files as the standard format for common multiplatform declarations metadata.
+> This change does not affect the main `kotlin-stdlib.jar` or `kotlin-stdlib-all.jar` artifacts.
+>
+> **Deprecation cycle**:
+>
+> - 2.1.0: deprecate and remove `kotlin-stdlib-common.jar` artifact
+
+### Deprecate appendln in favor of appendLine
+
+> **Issue**: [KTLC-27](https://youtrack.jetbrains.com/issue/KTLC-27)
+>
+> **Component**: kotlin-stdlib
+>
+> **Incompatible change type**: source
+>
+> **Short summary**: `StringBuilder.appendln()` is deprecated in favor of `StringBuilder.appendLine()`.
+>
+> **Deprecation cycle**:
+>
+> - 1.4.0: the `appendln()` function is deprecated; report a warning on use
+> - 2.1.0: raise the warning to an error
+
+### Deprecate freezing-related APIs in Kotlin/Native
+
+> **Issue**: [KT-69545](https://youtrack.jetbrains.com/issue/KT-NNNNN)
+>
+> **Component**: kotlin-stdlib
+>
+> **Incompatible change type**: source
+>
+> **Short summary**: Freezing-related APIs in Kotlin/Native, previously marked with the `@FreezingIsDeprecated` annotation, are now deprecated.
+> This aligns with the new memory manager that removes the need for freezing objects for thread sharing.
+> For migration details, see the [Kotlin/Native migration guide](native-migration-guide.md#update-your-code).
+>
+> **Deprecation cycle**:
+>
+> - 1.7.20: deprecate freezing-related APIs with a warning
+> - 2.1.0: raise the warning to an error
+
+### Change Map.Entry behavior to fail-fast on structural modification
+
+> **Issue**: [KTLC-23](https://youtrack.jetbrains.com/issue/KTLC-23)
+>
+> **Component**: kotlin-stdlib
+>
+> **Incompatible change type**: behavioral
+>
+> **Short summary**: Accessing a `Map.Entry` key-value pair after its associated map has been structurally modified now throws a `ConcurrentModificationException`.
+>
+> **Deprecation cycle**:
+>
+> - 2.1.0: throw an exception when a map structural modification is detected
+
+## Tools
+
+### Deprecate KotlinCompilationOutput#resourcesDirProvider
+
+> **Issue**: [KT-69255](https://youtrack.jetbrains.com/issue/KT-69255)
+>
+> **Component**: Gradle
+>
+> **Incompatible change type**: source
+>
+> **Short summary**: The `KotlinCompilationOutput#resourcesDirProvider` field is deprecated.
```

Comment:
Is there more context we could add here? For example:

The KotlinCompilationOutput#resourcesDirProvider field [FROM THE ...] is deprecated. Use KotlinSourceSet.resources [IN YOUR] to add additional resource directories instead.

Reply:
thank you 👍 I clarified that it's something in the build script!